lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+        (p. 6) for further explanation.
+    unroll_scan : bool
+        If True the recursion is unrolled instead of using scan. For some
+        graphs this gives a significant speed up but it might also consume
+        more memory. When `unroll_scan` is true then the `gradient_steps`
+        setting is ignored. The input sequence length cannot be specified as
+        None when `unroll_scan` is True.
+    precompute_input : bool
+        If True, precompute input_to_hid before iterating through
+        the sequence. This can result in a speedup at the expense of
+        an increase in memory usage.
+
+    References
+    ----------
+    .. [1] Graves, Alex: "Generating sequences with recurrent neural networks."
+           arXiv preprint arXiv:1308.0850 (2013).
+    """
+    def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
+                 nonlinearity=nonlinearities.rectify,
+                 hid_init=init.Constant(0.),
+                 backwards=False,
+                 learn_init=False,
+                 gradient_steps=-1,
+                 grad_clipping=False,
+                 unroll_scan=False,
+                 precompute_input=True,
+                 **kwargs):
+
+        super(CustomRecurrentLayer, self).__init__(incoming, **kwargs)
+
+        self.input_to_hidden = input_to_hidden
+        self.hidden_to_hidden = hidden_to_hidden
+        self.learn_init = learn_init
+        self.backwards = backwards
+        self.gradient_steps = gradient_steps
+        self.grad_clipping = grad_clipping
+        self.unroll_scan = unroll_scan
+        self.precompute_input = precompute_input
+
+        if unroll_scan and gradient_steps != -1:
+            raise ValueError(
+                "Gradient steps must be -1 when unroll_scan is true.")
+
+        if unroll_scan and self.input_shape[1] is None:
+            raise ValueError("Input sequence length cannot be specified as "
+                             "None when unroll_scan is True")
+
+        # Check that output shapes match
+        if input_to_hidden.output_shape != hidden_to_hidden.output_shape:
+            raise ValueError("The output shape for input_to_hidden and "
+                             "input_to_hidden must be equal, but "
+                             "input_to_hidden.output_shape={} and "
+                             "hidden_to_hidden.output_shape={}".format(
+                                 input_to_hidden.output_shape,
+                                 hidden_to_hidden.output_shape))
+
+        if nonlinearity is None:
+            self.nonlinearity = nonlinearities.identity
+        else:
+            self.nonlinearity = nonlinearity
+
+        # Get the input dimensionality and number of units based on the
+        # expected output of the input-to-hidden layer
+        self.num_inputs = np.prod(self.input_shape[2:])
+        self.num_units = input_to_hidden.output_shape[-1]
```

Comment:
So, here's a fork where `CustomRecurrentLayer` is changed so that it doesn't flatten trailing feature dimensions and can work fine with convolutional input-to-hidden/hidden-to-hidden connections: https://github.com/craffel/nntools/commit/e9f44fc0c6766c2ffd4668b27545a27aa86cde31

e.g.:
```Python
n_batch, n_steps, n_channels, width, height = (2, 3, 4, 5, 6)
n_filters = 7
filter_size = (3, 3)

l_in = lasagne.layers.InputLayer((n_batch, n_steps, n_channels, width, height))
l_in_hid = lasagne.layers.Conv2DLayer(
    lasagne.layers.InputLayer((n_batch, n_channels, width, height)),
    n_filters, filter_size, border_mode='same')
l_hid_hid = lasagne.layers.Conv2DLayer(
    lasagne.layers.InputLayer((n_batch, n_filters, width, height)),
    n_filters, filter_size, border_mode='same')
l_rec = CustomRecurrentLayer(l_in, l_in_hid, l_hid_hid)
out = lasagne.layers.get_output(l_rec)
test_in = np.zeros((n_batch, n_steps, n_channels, width, height), np.float32)
print out.eval({l_in.input_var: test_in}).shape
```
yields `(2, 3, 7, 5, 6)`.  Note that `border_mode` has to be `same`, because otherwise on each pass of `step` the hidden state shrinks!  This is what was tripping me up for a little while yesterday...  Anyways, we can wait to merge this in, as it warrants more testing and usage (although I don't know of anyone actually doing stuff like this in practice).